File: ruby/sum-of-multiples/sum_of_multiples.rb
require 'set'

class SumOfMultiples
  attr_reader :factors

  def initialize(*factors)
    @factors = Set.new(factors)
  end

  def to(limit)
    factors.flat_map do |factor|
      return 0 if factor.zero?

      1.step
       .lazy
       .map { |i| i * factor }
       .take_while { |i| i < limit }
       .force
    end.uniq.sum
  end
end

module BookKeeping
  VERSION = 2
end

File: scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sum(factors: Set[Int], limit: Int): Int =
    factors.size match {
      case 0 => 0
      case _ =>
        (factors.min until limit)
          .filter(n => factors.exists(n % _ == 0))
          .sum
    }
}